Separate the csv adapter from the query building logic

dlindhol opened the pull request:
This depends on the latis3 dataset-as-trait branch (which has been recently rebased on master) but the key elements are here. 

I've separated the query building logic into the `HapiAdapter`, an abstract class whose subclasses must provide the dataset format string (for the query: csv, json, binary) and an Adapter to use for parsing the data. The `HapiCsvAdapter` extends it to specify "csv" and uses the TextAdapter.

We should be able to refactor the HAPI JSON adapter in a similar way where it then becomes the `parsingAdapter`.

dlindhol commented on src/test/scala/latis/input/HapiAdapterSpec.scala:
```diff
@@ -0,0 +1,57 @@
+    adapter,
+    baseUri
+  )
+//  val dataset = Dataset.fromName("sorce_tsi") //get from fdml
```
Any thoughts on defining the dataset in code vs using fdml?

sapols replied:
Sorry, I haven't actually used an FDML before. What would it look like defined in FDML?

dlindhol replied:
Very similar to tsml: https://github.com/latis-data/latis3/blob/master/core/src/test/resources/datasets/data.fdml
I'm just not sure if we want a bunch of test fdml files floating around or to define test datasets in code closer to the test.

sapols replied:
If it's just for the tests then I lean toward defining the dataset in code. That way when I'm going to look at a test, I don't also have to look in an FDML to see the dataset.

sapols replied:
If it's just for the tests then I lean toward defining the dataset in code. That way when I'm going to look at a test, I don't also have to look in an FDML to see the dataset.

dlindhol commented on src/test/scala/latis/input/HapiAdapterSpec.scala:
```diff
@@ -0,0 +1,57 @@
+  
+  
+  "A hapi csv request with time selections" should "return csv records" in {
+    val ds = dataset.withOperation(Selection("time", ">", "2010-01-01"))
```
Note the new `withOperation` that the new dataset stuff supports. The DSL could add a `select` method on `Dataset`.

dlindhol commented on src/test/scala/latis/input/HapiAdapterSpec.scala:
```diff
@@ -0,0 +1,57 @@
+    val ds = dataset.withOperation(Selection("time", ">", "2010-01-01"))
+                    .withOperation(Selection("time", "<", "2011-01-01"))
+    
+    StreamUtils.unsafeHead(ds.samples) match {
```
Always looking for convenience methods to make life easier. Maybe we could import a test DSL that adds unsafe things to `Dataset` like `head` that would immediately give us the first `Sample`?

sapols commented on src/main/scala/latis/input/HapiAdapter.scala:
```diff
@@ -4,40 +4,103 @@ import latis.model._
+      case _ => ??? //TODO: bug, we should only be given Ops that pass canHandleOperation
+    }
+    //TODO: if we don't have both, use defaults from info header
+    //TODO: add epsilon if times are the same? (CDAWeb seems to require it)
```
I do think adding an epsilon would be helpful.

lindholc commented on src/main/scala/latis/input/HapiAdapter.scala:
```diff
@@ -4,40 +4,103 @@ import latis.model._
+            throw new UnsupportedOperationException(msg)
+        }
+      //case Projection(vids @ _*) => vids.mkString("parameters=", ",", "")
+      case _ => ??? //TODO: bug, we should only be given Ops that pass canHandleOperation
```
Should we throw an exception here?

lindholc commented on src/main/scala/latis/input/HapiAdapter.scala:
```diff
@@ -4,40 +4,103 @@ import latis.model._
+    var startTime: Option[String] = None
+    
+    // Placeholder for max time
+    var endTime:   Option[String] = None
```
These variables don't appear to be used anywhere.

dlindhol replied:
These will be used to support the TODOs, to which I should probably do.

lindholc commented on src/main/scala/latis/input/HapiAdapter.scala:
```diff
@@ -4,40 +4,103 @@ import latis.model._
-import latis.data.HapiFunction
+import latis.ops.Operation
+import latis.ops.Selection
+import latis.ops.Projection
```
This import isn't used.

dlindhol replied:
Hmm, eclipse is no longer warning me of unused imports. Now if we can figure out how to get IDEs and/or scalafmt to do this for us....

lindholc commented on src/test/scala/latis/input/HapiAdapterSpec.scala:
```diff
@@ -0,0 +1,57 @@
+  
+  val baseUri = new URI("http://lasp.colorado.edu/lisird/hapi/")
+  
+  val dataset = new AdaptedDataset(
```
To reduce namespace clutter, you could make the dataset like this:

```scala
val dataset = {
  val model = ???
  val adapter = ???
  val baseUri = ???

  new AdaptedDataset(...)
}
```

dlindhol commented:
I handled some of the TODOs and cleaned up a few things. The only bit I'd like to get in here for this PR is the use of time coverage from the HAPI "info" for default time selections. I might borrow something from Chris' HapiReader branch. Projection push down might be nice but we can live without that.

I also changed the test to use a dataset that actually works.
